src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -404,12 +411,106 @@ void FancyZones::MoveWindowIntoZone(HWND window, winrt::com_ptr<IZoneWindow> zon
+    if (GetWindowPlacement(window, &placement))
+    {
+        return IsSameRect(expected, placement.rcNormalPosition);
+    }
+    return false;
+}
+
+RECT FitOnScreen(const RECT& windowRect, const RECT& originMonitorRect, const RECT& destMonitorRect)
+{
+    // New window position on active monitor. If window fits the screen, this will be final position.
+    int left = destMonitorRect.left + (windowRect.left - originMonitorRect.left);
+    int top = destMonitorRect.top + (windowRect.top - originMonitorRect.top);
+    int W = RectWidth(windowRect);
+    int H = RectHeight(windowRect);
+
+    if ((left < destMonitorRect.left) || (left + W > destMonitorRect.right))
+    {
+        // Set left window border to left border of screen (add padding). Resize window width if needed.
+        left = destMonitorRect.left + LEFT_TOP_PADDING;
+        W = min(W, RectWidth(destMonitorRect) - LEFT_TOP_PADDING);
+    }
+    if ((top < destMonitorRect.top) || (top + H > destMonitorRect.bottom))
+    {
+        // Set top window border to top border of screen (add padding). Resize window height if needed.
+        top = destMonitorRect.top + LEFT_TOP_PADDING;
+        H = min(H, RectHeight(destMonitorRect) - LEFT_TOP_PADDING);
+    }
+
+    return { left, top, left + W, top + H };
+}
+
+void OpenWindowOnActiveMonitor(HWND window, HMONITOR monitor) noexcept
+{
+    // By default Windows opens new window on primary monitor.
+    // Try to preserve window width and height, adjust top-left corner if needed.
+    HMONITOR origin = MonitorFromWindow(window, MONITOR_DEFAULTTOPRIMARY);
+    if (origin == monitor)
+    {
+        // Certain applications by design open in last known position, regardless of FancyZones.
+        // If that position is on currently active monitor, skip custom positioning.
+        return;
+    }
+
+    WINDOWPLACEMENT placement{};
+    if (GetWindowPlacement(window, &placement))
+    {
+        MONITORINFOEX originMi;
+        originMi.cbSize = sizeof(originMi);
+        if (GetMonitorInfo(origin, &originMi))
+        {
+            MONITORINFOEX destMi;
+            destMi.cbSize = sizeof(destMi);
+            if (GetMonitorInfo(monitor, &destMi))
+            {
+                RECT newPosition = FitOnScreen(placement.rcNormalPosition, originMi.rcWork, destMi.rcWork);
+                SizeWindowToRect(window, newPosition);
+
+                // Certain applications could perform their own custom positioning afterwards, so check if window
+                // is in expected position, and if not, try to move it there again.
+                for (int i = 0; i < CUSTOM_POSITIONING_MAX_RETRIES; ++i)
+                {
+                    std::this_thread::sleep_for(std::chrono::milliseconds(CUSTOM_POSITIONING_WAIT_TIME));
+                    if (WindowInExpectedPosition(window, newPosition))
+                    {
+                        break;
+                    }
+                    SizeWindowToRect(window, newPosition);

```

Comment:
We should discuss whether we need this resizing. If it does happen, it won't look pretty.

Reply:
This is something that I've discussed with @enricogior for some time, and it is partially explained in comment above. For example, Firefox has it's own custom positioning implementation, and sometimes first call to position Firefox does not give expected result, Firefox stays in it's own position. This could be confusing to our customers, so rather with a price of small glitch (which is something that is not that rare when doing magic with displays in Windows OS in general) we keep our functionality more robust and same for all applications.

Reply:
Okay, that sounds reasonable